src/metaopt/io/database/mongodb.py

```diff
@@ -0,0 +1,277 @@
+        selection : dict, optional
+           Used in `atomic` update operation: elements of the matched entry to
+           return, the projection.
+
+        :return: operation success, if not `atomic`, matched updated document else.
+
+        .. note::
+           In the case of an insert operation, `data` variable will be updated
+           to contain a unique *_id* key.
+
+        .. note::
+           In the case of an update operation, if `query` fails to find a
+           document that matches, insert of `data` will be performed instead.
+        """
+        dbcollection = self._db[collection_name]
+
+        if query is None:
+            # We can assume that we do not want to update.
+            # So we do insert_many instead.
+            if type(data) not in (list, tuple):
+                data = [data]
+            result = dbcollection.insert_many(documents=data)
+            return result.acknowledged
+
+        update_data = {'$set': data}
+        if atomic is False:
+            result = dbcollection.update_many(filter=query,
+                                              update=update_data,
+                                              upsert=True)
+            return result.acknowledged
+        elif atomic is True:
+            selected_one_key, selection = _get_selected_one_key(selection)
+            result = dbcollection.find_one_and_update(filter=query,
+                                                      update=update_data,
+                                                      projection=selection,
+                                                      upsert=True,
+                                                      sort=[('$natural', -1)],
+                                                      return_document=ReturnDocument.AFTER)
+            return _get_leaf_value(selected_one_key, result) if selected_one_key else result
+
+    def read(self, collection_name, query, selection=None, atomic=False):
+        """Read a collection and return a value according to the query.
+
+        Arguments
+        ---------
+        collection_name : str
+           A collection inside database, a table.
+        query : dict
+           Filter entries in collection.
+        selection : dict, optional
+           Elements of matched entries to return, the projection.
+        atomic : bool, optional
+           If True, return only the most recent entry that matches `query`.
+
+        :return: matched document[s]
+        """
+        dbcollection = self._db[collection_name]
+
+        selected_one_key, selection = _get_selected_one_key(selection)
+
+        if atomic is False:
+            cursor = dbcollection.find(query, selection)
+            dbdocs = list(cursor)
+        elif atomic is True:
+            # Return the most recent of the query
+            dbdocs = list(dbcollection.find(query, selection, sort=[('$natural', -1)], limit=1))
```

Comment:
In this case `atomic` means that the operation is safe for concurrency. The atomicity is about the operation. The operation could return many documents.

With the current implementation we are force to return a single document in `write()` because of MongoDB particularities, but in general `write()` could be applied on many documents _atomically_. We are forcing `read()`s behavior to stay coherent on `write()`s behavior, while the later was already forced because of a specific database particularity. That doesn't sound like a good design to me.

It is my belief that we should give ourselves some time to figure out what would be the best way to integrate atomicity. Not having it won't hurt us for the MVP.

Reply:
I see your point. I think though that I will find useful searching for a single most recent document later on, for MVP. We could change the name of the optional argument though to something else. I believe that in general this is ok for now. `AbstractDB` is an internal API, we could change this later.

Reply:
I understand having a way to fetch the most recent document can be handy, but I disagree this is how it should be done. If I read the following code snippet, it's not clear why the output would be the most recent row unless I read the doc.
```python
most_recent_row = db.read(some_query, atomic=True)
```

The following is less elegant but more explicit
```python
most_recent_row = next(sorted(db.read(some_query), key=lambda d: d["date"]))
```

And could later be turned into
```python
most_recent_row = db.read(some_query, sort={"date": 1})
```

Anyhow, we don't need sort for now, and adding it would not impact the interface so we don't need to rush for it. However, adding it now would affect our progress. When we need it, we can simply make a PR for sort only. 

It might feel like I'm arguing for details and slowing down our progress as I advocate for removing stuff rather than changing or adding, but we need to focus more. The tendency of adding extra features will hurt us more and more as we progress with a larger and more complicated code base. Every extra feature is adding burden on development and maintenance. When the code is designed properly, extra features are easy to integrate later on, so there is no reason to rush them in. What we should rather rush for is users' feedback. We cannot afford to design hastily for that, but we can spare time on extra features.

Reply:
I would just like an optional interface to return the most return recent thing. Databases can do that generally. I would not like a general `sort` argument, this can be done as you demonstrated. However, searching for the most recent something is probably going to be common. I am not sure.

I agree with your last comment. It takes unjustified motivation to remove stuff I tried right now to facilitate for my future coding. However, I will try to contain my urge to implement in front stuff before they become necessary, in the future.

What would you suggest that I should do? Should I remove it? I think the current tests describe pretty much what I will need to proceed. Interface can be polished later, no need to perfect it right now as you said.

Reply:
> It takes unjustified motivation to remove stuff I tried right now to facilitate for my future coding. However, I will try to contain my urge to implement in front stuff before they become necessary, in the future.

I don't understand what you mean.

> What would you suggest that I should do? Should I remove it? I think the current tests describe pretty much what I will need to proceed. Interface can be polished later, no need to perfect it right now as you said.

I think the best would be to move the extra feature to a new branch and point to it in an issue. This way we do not loose trace of the code but do not have the burden of maintaining it. From what I see there should be 2 issues. One about atomicity and one about getting the most recent document only. Those two should not be coupled.